lib/httplib/csrf/csrf.go

```diff
@@ -0,0 +1,155 @@
+/*
+Copyright 2016 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package csrf
+
+import (
+	"crypto/rand"
+	"crypto/subtle"
+	"encoding/base64"
+	"net/http"
+
+	"github.com/gravitational/trace"
+)
+
+const (
+	// CookieName is a name of the cookie
+	CookieName = "grv_csrf"
+	// HeaderName is the default HTTP request header to inspect
+	HeaderName = "X-CSRF-Token"
+	// tokenLength is CSRF token length in bytes.
+	tokenLength = 32
+	// defaultMaxAge is the default MaxAge for cookies.
+	defaultMaxAge = 0
+	// invalidErrText is user error message about invalid token
+	invalidErrText = "failed to validate csrf token"
+)
+
+// Protect adds CSRF token into the user session via secure cookie,
+// it implements "double submit cookie" approach to check against csrf attacks
+// https://www.owasp.org/index.php/Cross-Site_Request_Forgery_%28CSRF%29_Prevention_Cheat_Sheet#Double_Submit_Cookie
+func Protect(w http.ResponseWriter, r *http.Request) (string, error) {
+	// retrieve the token from the session.
+	token, err := extractFromCookie(r)
+	// if there was an error retrieving the token, the token doesn't exist
+	if err != nil || len(token) != tokenLength {
```

Comment:
if the token length does not match, it's someone altering the token, so return error here.